File: Web/profile-client/src/lib/profiles/profile-reader/eoa-signature-verifier.ts
// src/lib/profiles/profile-reader/eoa-signature-verifier.ts
import { SigningKey, computeAddress, hexlify } from "ethers";
import { keccak_256 } from "@noble/hashes/sha3";
import { SECP_N_OVER_2 } from "$lib/profiles/consts";
import { bytesToBigInt } from "$lib/profiles/profile-reader/utils";
import type { Hex } from "$lib/profiles/profile-reader/utils";

/**
 * Verifies EOA ECDSA signatures over a 32-byte Keccak hash.
 * Enforces EIP-2 (low-S). No chain calls required.
 */
export class EoaSignatureVerifier {
    /**
     * Synchronous verification over a 32-byte hash.
     */
    verify(hash32: Uint8Array, signerAddress: Hex | string, signature65: Uint8Array): boolean {
        if (!(hash32 instanceof Uint8Array)) {
            throw new Error("hash must be a Uint8Array");
        }
        const is32 = hash32.length === 32;
        if (!is32) {
            throw new Error(`hash must be 32 bytes, got ${hash32.length}`);
        }

        if (!(signature65 instanceof Uint8Array)) {
            throw new Error("signature must be a Uint8Array");
        }
        const is65 = signature65.length === 65;
        if (!is65) {
            throw new Error(`signature must be 65 bytes, got ${signature65.length}`);
        }

        const addrEmpty = !signerAddress || String(signerAddress).trim().length === 0;
        if (addrEmpty) {
            throw new Error("Empty address");
        }

        // EIP-2: reject non-canonical S (and s != 0)
        const sBytes = signature65.subarray(32, 64);
        const sBI = bytesToBigInt(sBytes);
        const sIsZero = sBI === 0n;
        const sIsHigh = sBI > SECP_N_OVER_2;
        const sCanonical = !sIsZero && !sIsHigh;
        if (!sCanonical) {
            return false;
        }

        // Normalize v to {27,28}; accept {0,1} too (no other variants).
        const vRaw = signature65[64];
        const v = (vRaw === 27 || vRaw === 28) ? vRaw
            : (vRaw === 0 || vRaw === 1) ? vRaw + 27
                : -1;
        const vValid = v === 27 || v === 28;
        if (!vValid) {
            return false;
        }

        const r = hexlify(signature65.subarray(0, 32));
        const s = hexlify(signature65.subarray(32, 64));

        try {
            const pub = SigningKey.recoverPublicKey(hash32, { r, s, v });
            const recovered = computeAddress(pub);
            return recovered.toLowerCase() === String(signerAddress).toLowerCase();
        } catch {
            return false;
        }
    }

    /**
     * Convenience for readers that only have the canonical payload bytes.
     * EOAs sign the keccak of those bytes.
     */
    verifyOverBytes(payloadBytes: Uint8Array, signerAddress: Hex | string, signature65: Uint8Array): boolean {
        if (!(payloadBytes instanceof Uint8Array)) {
            throw new Error("payloadBytes must be a Uint8Array");
        }
        const h = keccak_256(payloadBytes);
        return this.verify(h, signerAddress, signature65);
    }

    /* ─────────────────── async aliases to fit existing call sites ─────────────────── */

    async verifyAsync(
        hash32: Uint8Array,
        signerAddress: Hex | string,
        signature65: Uint8Array,
        _signal: AbortSignal | undefined = undefined
    ): Promise<boolean> {
        return this.verify(hash32, signerAddress, signature65);
    }

    async verify1271WithBytesAsync(
        payloadBytes: Uint8Array,
        signerAddress: Hex | string,
        signature65: Uint8Array,
        _signal: AbortSignal | undefined = undefined
    ): Promise<boolean> {
        return this.verifyOverBytes(payloadBytes, signerAddress, signature65);
    }
}

File: Circles.Profiles.Sdk/EoaSignatureVerifier.cs
using Circles.Profiles.Sdk.Utils;
using Nethereum.Hex.HexConvertors.Extensions;
using Nethereum.Signer;

namespace Circles.Profiles.Sdk;

/// <summary>
/// Verifies EOA ECDSA signatures over a 32-byte Keccak hash.
/// Enforces EIP-2 (low-S). No chain calls required.
/// </summary>
public sealed class EoaSignatureVerifier
{
    /// <summary>
    /// Synchronous verification over a 32-byte hash.
    /// </summary>
    public bool Verify(byte[] hash, string signerAddress, byte[] signature)
    {
        ArgumentNullException.ThrowIfNull(hash);
        ArgumentNullException.ThrowIfNull(signature);

        if (string.IsNullOrWhiteSpace(signerAddress))
        {
            throw new ArgumentException("Empty address", nameof(signerAddress));
        }

        string sigHex = "0x" + signature.ToHex();
        var sig = EthECDSASignatureFactory.ExtractECDSASignature(sigHex);

        if (!sig.IsLowS) return false; // EIP-2

        string? recovered = EthECKey.RecoverFromSignature(sig, hash)?.GetPublicAddress();
        return recovered?.Equals(signerAddress, StringComparison.OrdinalIgnoreCase) == true;
    }

    /// <summary>
    /// Convenience for readers that only have the canonical payload bytes.
    /// EOAs sign the keccak of those bytes.
    /// </summary>
    public bool VerifyOverBytes(byte[] payloadBytes, string signerAddress, byte[] signature)
    {
        ArgumentNullException.ThrowIfNull(payloadBytes);
        ArgumentNullException.ThrowIfNull(signature);

        byte[] h = Sha3.Keccak256Bytes(payloadBytes);
        return Verify(h, signerAddress, signature);
    }

    /* ─────────────────── async aliases to fit existing call sites ─────────────────── */

    /// <summary>
    /// Async alias for <see cref="Verify"/> to match existing facade calls.
    /// </summary>
    public Task<bool> VerifyAsync(
        byte[] hash,
        string signerAddress,
        byte[] signature,
        CancellationToken _ = default)
        => Task.FromResult(Verify(hash, signerAddress, signature));

    /// <summary>
    /// Async alias for <see cref="VerifyOverBytes"/> to match existing facade calls.
    /// Named to be compatible with older DefaultSignatureVerifier helper usage.
    /// </summary>
    public Task<bool> Verify1271WithBytesAsync(
        byte[] payloadBytes,
        string signerAddress,
        byte[] signature,
        CancellationToken _ = default)
        => Task.FromResult(VerifyOverBytes(payloadBytes, signerAddress, signature));
}
